### PR TITLE
Kc 907 public api implementation

### PR DESCRIPTION
**Pull request by jpkeepersecurity**
# Enterprise API Keys Management Implementation

## 📋 **Overview**

This PR implements a comprehensive enterprise API keys management system for Keeper Commander, providing three core operations: **list**, **generate**, and **revoke** API keys. The implementation follows the existing codebase patterns and includes extensive testing coverage.

### **Core Operations**
- **`public-api-key list`** - Display all enterprise API keys with comprehensive filtering and output options
- **`public-api-key generate`** - Create new API keys with role-based permissions and flexible expiration settings  
- **`public-api-key revoke`** - Revoke existing API keys with confirmation workflow

### **Key Capabilities**
- ✅ **Multiple Output Formats**: Table (default), JSON, CSV with file export support
- ✅ **Flexible Expiration Options**: 24h, 7d, 30d, 1y, or never expires
- ✅ **Role-Based Access Control**: Support for SIEM, CSPM, BILLING roles with READ/READ_WRITE permissions
- ✅ **Status Detection**: Automatic expired vs active status based on expiration dates
- ✅ **Interactive Confirmations**: User-friendly prompts with force override options
- ✅ **Comprehensive Help System**: Detailed examples and usage guides for each operation

### **File Structure**
```
keepercommander/commands/enterprise_api_keys.py  # Main implementation
unit-tests/test_command_enterprise_api_keys.py    # Comprehensive test suite
```

### **Architecture**
- **`ApiKeyCommand`** - Main group command following `GroupCommand` pattern
- **`ApiKeyListCommand`** - Handles listing operations with filtering and formatting
- **`ApiKeyGenerateCommand`** - Manages API key creation with validation and role assignment
- **`ApiKeyRevokeCommand`** - Handles revocation with user confirmation workflows

### **Integration**
- Follows existing enterprise command patterns from `scim.py` and `enterprise.py`
- Integrates with existing REST API infrastructure via `communicate_rest()`
- Uses protobuf message types from `publicapi_pb2` module
- Registered in enterprise commands via `base.py` registration system

### **List API Keys**
```bash
# Basic table view
public-api-key list

# JSON output with file export  
public-api-key list --format json --output api_keys.json

# CSV export
public-api-key list --format csv --output api_keys.csv
```

### **Generate API Keys**
```bash
# Basic API key with 30-day expiration
public-api-key generate --name "SIEM Integration" --roles "SIEM:2" --expires 30d

# Multiple roles with different permissions
public-api-key generate --name "Multi-Role Key" --roles "SIEM:2,CSPM:1,BILLING:2" --expires 1y

# Permanent API key (never expires)
public-api-key generate --name "Permanent Tool" --roles "SIEM:1" --expires never

# JSON output to file
public-api-key generate --name "Backup Tool" --roles "CSPM:2" --format json --output backup_key.json
```

### **Revoke API Keys**
```bash
# Interactive revocation with confirmation
public-api-key revoke 12345

# Force revocation without confirmation
public-api-key revoke 12345 --force
```

### **Supported Roles**
- **SIEM** (ID: 1) - Security Information and Event Management
- **CSPM** (ID: 2) - Cloud Security Posture Management  
- **BILLING** (ID: 3) - Billing and usage management

### **Action Types**
- **1 = READ** - Read-only access
- **2 = READ_WRITE** - Full read/write access

### **Expiration Options**
- **24h** - 24 hours from creation
- **7d** - 7 days from creation
- **30d** - 30 days from creation
- **1y** - 1 year from creation
- **never** - No expiration (permanent access)

### **Test Coverage Statistics**
- **Total Tests**: 27 comprehensive test scenarios
- **Success Rate**: 100% (all tests passing)
- **Coverage Areas**: All operations, error handling, edge cases, output formats

#### **📋 List Operations (8 tests)**
- `test_api_key_list_success` - Basic table format output validation
- `test_api_key_list_json_format` - JSON output structure validation  
- `test_api_key_list_json_comprehensive_fields` - Complete field validation with expected data
- `test_api_key_main_command_default_list_behavior` - Default command behavior
- `test_api_key_status_detection_expired_vs_active` - Status logic validation

#### **🔑 Generate Operations (10 tests)**
- `test_api_key_generate_success_matching_terminal_example` - Exact Commander output matching
- `test_api_key_generate_success_7d_expiration` - 7-day expiration validation
- `test_api_key_generate_success_30d_expiration` - 30-day expiration validation
- `test_api_key_generate_success_1y_expiration` - 1-year expiration validation
- `test_api_key_generate_success_never_expires` - Never expires validation
- `test_api_key_generate_multiple_roles` - Multi-role support
- `test_api_key_generate_multiple_roles_comprehensive` - Complex role combinations
- `test_api_key_generate_json_output` - JSON format validation
- `test_api_key_generate_json_comprehensive_fields` - Complete JSON structure validation
- `test_api_key_generate_json_with_output_file` - File export functionality

#### **🗑️ Revoke Operations (5 tests)**
- `test_api_key_revoke_matching_terminal_example` - Exact Commander output matching
- `test_api_key_revoke_cancelled_by_user` - User cancellation workflow
- `test_api_key_revoke_force_flag` - Force revocation without confirmation
- `test_api_key_revoke_success` - Basic revocation success

#### **❌ Error Handling (4 tests)**
- `test_api_key_generate_missing_name` - Required parameter validation
- `test_api_key_generate_missing_roles` - Role requirement validation
- `test_api_key_generate_invalid_role_format` - Input format validation
- `test_api_key_generate_invalid_role_name` - Role name validation
- `test_api_key_revoke_missing_token_id` - Token ID requirement validation
- `test_api_key_revoke_invalid_token_id` - Token ID format validation

### **Mock Data Validation**

The test suite uses **realistic mock data** that matches actual Commander terminal output:

```python
# Sample tokens matching real terminal examples
Token ID: 43, 44, 45 (Expired) - "Token Test", "Token Test 2"
Token ID: 53, 54 (Active) - "CSPM Tool", "Token For My Tests 111"
Enterprise ID: 8560
Roles: "SIEM:2, CSPM:2", "SIEM:2, CSPM:1", "SIEM:2, CSPM:2, Billing:2"
```

### **Field Validation**

Each test validates **all output fields**:
- ✅ Token ID, Enterprise ID, Name, Status
- ✅ Issued Date, Expiration Date formatting  
- ✅ Role assignments and action types
- ✅ Status logic (Active/Expired/Never)
- ✅ JSON structure completeness
- ✅ File I/O operations

## 🔄 **Command Name Evolution**

Initially implemented as `api-key`, the command was renamed to `public-api-key` for better specificity:

### **Updated References**
- ✅ Command registration: `commands['public-api-key']`
- ✅ All argument parsers: `prog='public-api-key <operation>'`
- ✅ Help documentation and examples
- ✅ Test suite assertions
- ✅ Interactive help system

### **Follows Existing Patterns**
- ✅ **Command Structure**: Mirrors `scim.py` implementation patterns
- ✅ **Error Handling**: Consistent with `enterprise.py` standards
- ✅ **REST API Integration**: Uses existing `communicate_rest()` infrastructure
- ✅ **Argument Parsing**: Follows Commander argparse conventions
- ✅ **Output Formatting**: Leverages existing `dump_report_data()` utilities

### **Enterprise Integration**
- ✅ Inherits from `EnterpriseCommand` base class
- ✅ Registered in `register_enterprise_commands()` function
- ✅ Uses enterprise protobuf message types
- ✅ Follows enterprise authentication patterns

### **Mock System**
- **REST API Mocking**: Complete `communicate_rest()` simulation
- **Protobuf Integration**: Realistic `publicapi_pb2` message generation
- **File I/O Testing**: Temporary file creation and cleanup
- **User Interaction**: Input/confirmation workflow simulation

### **Test Execution**
```bash
# Run all enterprise API key tests
python -m pytest unit-tests/test_command_enterprise_api_keys.py -v

# Results: 27 passed, 1 warning in 2.48s
```

## 🚀 **Ready for Production**

This implementation provides:
- ✅ **Complete Functionality**: All three core operations fully implemented
- ✅ **Robust Testing**: 27 comprehensive test scenarios with 100% pass rate
- ✅ **Error Handling**: Comprehensive validation and user-friendly error messages
- ✅ **Documentation**: Extensive help system with practical examples
- ✅ **Integration**: Seamless integration with existing Commander infrastructure
- ✅ **Code Quality**: Follows established patterns and best practices

The enterprise API keys management system is ready for immediate use with full confidence in its reliability and maintainability.

### CODE REVIEW

File: keepercommander/commands/enterprise_api_keys.py [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+
+
+class ApiKeyRevokeCommand(EnterpriseCommand):
+    def get_parser(self):
+        return api_key_revoke_parser
+
+    def execute(self, params, **kwargs):
+        token_id = kwargs.get('token_id')
+        if not token_id:
+            print("Token ID is required")
```
Comment (sk-keeper) [inline keepercommander/commands/enterprise_api_keys.py:378]: It looks like this line can be replaced with 
raise CommandError("Token ID is required")

Generally the Commander code uses `print` function to intentionally write some information to stdout.
In most cases it targets using commander in customer's scripts. Any other use of `print` is suspicious.

Any user information, especially warnings and errors, will generally use logging.*

raise CommandError is used to abnormally stop any command. This exception expects the message is clear and customer can correct command input.

Any other exception is not expected. These exceptions show standard message to switch to `debug` mode and repeat the command.
`debug` mode prints call stack that helps developers to identify the crash.

**Review by sk-keeper — APPROVED** [review]
